test/proxy/OwnedProxyWithReference.test.ts

```diff
@@ -0,0 +1,103 @@
+    tusd = await new MockTrueCurrencyFactory(owner).deploy()
+    implementationReference = await new ImplementationReferenceFactory(owner).deploy(tusd.address)
+  })
+
+  describe('Ownership', () => {
+    it('owner is the owner of the proxy', async () => {
+      expect(await proxy.proxyOwner()).to.equal(owner.address)
+    })
+
+    it('owner can transfer proxy ownership ', async () => {
+      expect(await proxy.pendingProxyOwner()).to.equal(ZERO_ADDRESS)
+      await proxy.transferProxyOwnership(anotherWallet.address)
+
+      expect(await proxy.pendingProxyOwner()).to.equal(anotherWallet.address)
+    })
+
+    it('pending owner can claim ownership ', async () => {
+      await proxy.transferProxyOwnership(anotherWallet.address)
+      await proxy.connect(anotherWallet).claimProxyOwnership()
+
+      expect(await proxy.proxyOwner()).to.equal(anotherWallet.address)
+    })
+
+    it('non owner cannot transfer ownership ', async () => {
+      await expect(proxy.connect(anotherWallet).transferProxyOwnership(anotherWallet.address))
+        .to.be.reverted
+    })
+
+    it('non pending owner cannot claim ownership ', async () => {
+      await proxy.transferProxyOwnership(anotherWallet.address)
+      await expect(proxy.connect(thirdWallet).claimProxyOwnership())
+        .to.be.reverted
+    })
+
+    it('zero address cannot be pending owner ', async () => {
+      await expect(proxy.transferProxyOwnership(ZERO_ADDRESS))
+        .to.be.reverted
+    })
+
+    it('emits proper events', async () => {
+      await expect(proxy.transferProxyOwnership(anotherWallet.address))
+        .to.emit(proxy, 'NewPendingOwner')
+        .withArgs(owner.address, anotherWallet.address)
+      await expect(proxy.connect(anotherWallet).claimProxyOwnership())
+        .to.emit(proxy, 'ProxyOwnershipTransferred')
+        .withArgs(owner.address, anotherWallet.address)
+    })
+  })
+
+  describe('Referencing', () => {
+    it('sets up implementation reference ', async () => {
+      await proxy.changeImplementationReference(implementationReference.address)
+      expect(await proxy.implementation()).to.equal(tusd.address)
+    })
+
+    it('non owner cannot upgrade implementation contract', async () => {
+      await expect(proxy.connect(anotherWallet).changeImplementationReference(implementationReference.address))
+        .to.be.reverted
+    })
+
+    it('emits proper event', async () => {
+      await expect(proxy.changeImplementationReference(implementationReference.address))
+        .to.emit(proxy, 'ImplementationReferenceChanged')
+        .withArgs(implementationReference.address)
+    })
+  })
+
+  it('calls implementation function', async () => {
+    await proxy.changeImplementationReference(implementationReference.address)
+    await tusd.attach(proxy.address).approve(owner.address, 1)
+    expect(await tusd.attach(proxy.address).allowance(owner.address, owner.address)).to.eq(1)
+  })
```

Comment:
Could you add tests with a function that returns/accepts a long string (over 256 bytes). There's a small chance that this could be broken

Reply:
This function would not accept anything, that is not a valid address.

Reply:
Not this, a different one. You'll have to write a contract with a simple function, e.g. that takes string and returns itself. And pass a long string there